packages/backend-core/src/timers/timers.ts

```diff
@@ -20,3 +20,39 @@ export function cleanup() {
   }
   intervals = []
 }
+
+export class ExecutionTimeoutError extends Error {
+  public readonly name = "ExecutionTimeoutError"
+}
+
+export class ExecutionTimeTracker {
+  static withLimit(limitMs: number) {
+    return new ExecutionTimeTracker(limitMs)
+  }
+
+  constructor(private limitMs: number) {}
+
+  private totalTimeMs = 0
+
+  track<T>(f: () => T): T {
+    this.checkLimit()
+    const [startSeconds, startNanoseconds] = process.hrtime()
+    const startMs = startSeconds * 1000 + startNanoseconds / 1e6
```

Comment:
You could use [`process.hrtime.bigint()`](https://nodejs.org/docs/latest-v18.x/api/process.html#processhrtimebigint) to get this calculation instead